addons/sale_timesheet/models/account.py

```diff
@@ -1,28 +1,42 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from odoo import api, models
+from odoo.exceptions import UserError
+from odoo import api, fields, models, _
 
 
 class AccountAnalyticLine(models.Model):
     _inherit = 'account.analytic.line'
 
+    timesheet_invoice_type = fields.Selection([
+        ('billable_time', 'Billable Time'),
+        ('billable_fixed', 'Billable Fixed'),
+        ('non_billable', 'Non Billable')], string="Billable Type", readonly=True, copy=False)
+    timesheet_invoice_id = fields.Many2one('account.invoice', string="Invoice", readonly=True, copy=False, help="Invoice created from the timesheet")
+    timesheet_revenue = fields.Monetary("Revenue", default=0.0, readonly=True, currency_field='company_currency_id', copy=False)
+
     @api.model
     def create(self, values):
         if values.get('task_id'):
             task = self.env['project.task'].browse(values['task_id'])
             values['so_line'] = task.sale_line_id.id or values.get('so_line', False)
-        values.update(self._get_timesheet_cost(values))
+        values.update(self._get_timesheet_values(values))
+        values.update(self._get_timesheet_billing_values(values))
         return super(AccountAnalyticLine, self).create(values)
 
     @api.multi
     def write(self, values):
+        # prevent to update invoiced timesheets
+        if self.filtered(lambda timesheet: timesheet.timesheet_invoice_id):
```

Comment:
I don"t understand. Do you means removing entries (amount, sol, ...) from the values dict ?